File: python/examples/fuzzing.py
#!/usr/bin/env python

import collections
import getopt
import random
import sys

from sdk.API import API
from sdk.data.CANFrame import CANFrame
from sdk.data.CANResponseFilter import CANResponseFilter
from sdk.data.Request import Request

fuzz_count = 10  # type: int
fuzz_ids = range(0, 0x700)


def count_stats(responses):
    """
    :param iterator of CANResponseIterator responses:
    :return:
    """
    ids = {}
    for response in responses:
        log_frame(response.request)
        for frame in response.iterator():
            log_frame(frame)
            if frame.frame_id in ids:
                ids[frame.frame_id] += 1
            else:
                ids[frame.frame_id] = 1
    print("Amount of unique IDs: " + str(len(ids.keys())))
    oids = collections.OrderedDict(sorted(ids.items()))
    for key, val in oids.items():
        print("ID: " + hex(key) + " was encountered " + str(val) + " time(s)")


def usage():
    print("Usage: fuzzing.py -k <key> -s <secret> [-d <\"device id\">]")
    sys.exit(2)


def setup(api_key, secret, device_id, cert_path):
    # create api connection
    api = API(cert_path=cert_path)
    # get authentication token
    token = api.authenticate(api_key, secret)
    # get device list from cloud
    devices = api.get_connected_devices(token)
    # find available device
    for dev in devices:
        if dev.is_available_now() and ((not device_id) or dev.device_id == device_id):
            print('Found device: ' + str(dev))
            return api.connect_to_device(token, dev)
    else:
        print("Couldn't find device")
        sys.exit(3)


def generate_request():
    # Generate random request to device
    data_length = random.randint(1, 7)
    data = bytearray((random.getrandbits(8) for i in range(data_length)))
    frame_id = fuzz_ids[random.randint(0, len(fuzz_ids) - 1)]
    can_frame = CANFrame(frame_id, data)
    request = Request(can_frame, 100)
    return request


def fuzz(api_key, secret, device_id, cert_path):
    cloud = setup(api_key, secret, device_id, cert_path)
    try:
        requests = [generate_request() for _ in range(fuzz_count)]
        ids = [hex(r.frame.frame_id) for r in requests]
        print("Send request with ids: " + str(ids))
        count_stats(cloud.send_can_frames(requests, CANResponseFilter.filter_min_max_ids(0, 0x300)))
    finally:
        cloud.close()


def log_frame(frame):
    if frame.response:
        print('<===' + frame.to_json())
    else:
        print('===>' + frame.to_json())


def main(argv):
    # Parsing CLI arguments
    api_key = ''
    secret = ''
    device_id = ''
    cert_path = ''
    try:
        opts, args = getopt.getopt(argv, "th:p:k:s:d:a:")
        for opt, arg in opts:
            if opt == '-k':
                api_key = arg
            elif opt == '-s':
                secret = arg
            elif opt == '-d':
                device_id = arg
            elif opt == '-a':
                cert_path = arg
        if api_key and secret:
            # starting real work
            fuzz(api_key, secret, device_id, cert_path)
        else:
            usage()
    except getopt.GetoptError:
        usage()


if __name__ == "__main__":
    try:
        main(sys.argv[1:])
    except Exception as e:
        print('Error: ')
        print(str(e))

File: java/examples/src/main/java/Fuzzing.java
import com.visualthreat.api.API;
import com.visualthreat.api.VTCloud;
import com.visualthreat.api.data.*;
import com.visualthreat.api.v1.Token;
import lombok.extern.slf4j.Slf4j;
import org.apache.commons.cli.CommandLine;
import org.apache.commons.cli.CommandLineParser;
import org.apache.commons.cli.DefaultParser;
import org.apache.commons.cli.Options;

import java.security.SecureRandom;
import java.util.*;
import java.util.stream.Collectors;
import java.util.stream.IntStream;

@Slf4j
public final class Fuzzing {
  // CLI arguments parsing init
  private static final Options options = new Options();
  private static final Random rnd = new SecureRandom();
  private static final List<CANFrame> frames = new LinkedList<>();
  private static final long WAIT_TIME = 100;
  // Amount of requests to send
  private static final int maxCount = 1000;
  private static VTCloud cloud = null;
  private static int count = 0;

  static {
    options.addOption(Arguments.DEVICE.type, "device", true, "Device ID");
    options.addOption(Arguments.KEY.type, "key", true, "API key");
    options.addOption(Arguments.SECRET.type, "secret", true, "Secret");
  }

  private static void showUsage() {
    log.error("Usage: -k <key> -s <secret> [-d <\"device id\">]");
    die();
  }

  private static void die() {
    if (cloud != null) {
      cloud.close();
    }
    System.exit(3);
  }

  public static void main(final String[] args) throws Exception {
    // parsing CLI arguments
    final CommandLineParser parser = new DefaultParser();
    final CommandLine cmd = parser.parse(options, args);
    if (cmd.hasOption(Arguments.KEY.type) &&
        cmd.hasOption(Arguments.SECRET.type)) {
      final String key = cmd.getOptionValue(Arguments.KEY.type);
      final String secret = cmd.getOptionValue(Arguments.SECRET.type);
      final String deviceId = cmd.hasOption(Arguments.DEVICE.type) ?
          cmd.getOptionValue(Arguments.DEVICE.type) : null;
      // Starting real work
      start(key, secret, deviceId);
    } else {
      showUsage();
    }
  }

  /**
   * Generating random Request, with random data, data length and random id.
   */
  private static Request generateRequest() {
    final int frameId = rnd.nextInt(0x800);
    final int dataLength = rnd.nextInt(8) + 1;
    final byte[] frameData = new byte[dataLength];
    rnd.nextBytes(frameData);
    return new Request(new CANFrame(frameId, frameData), WAIT_TIME);
  }

  private static void makeRequest(final VTCloud cloud) {
    log.info("Iteration # " + count);
    try {
      // Send requests to device and wait for a response
      final Collection<Request> requests = IntStream.range(0, 100)
          .mapToObj(i -> generateRequest())
          .collect(Collectors.toList());
      final Iterator<Response> responses = cloud.sendCANFrames(
          Collections.singletonList(generateRequest()), CANResponseFilter.filterIds(0, 0x300));

      while (responses.hasNext()) {
        final Response response = responses.next();
        handleResponse(response.getResponses());
      }
    } catch (final Exception e) {
      log.error("Something went wrong", e);
      die();
    }
  }

  /**
   * Counting simple stats from all responses.
   * Count unique ids, and count of concrete id.
   */
  private static void handleData(final List<CANFrame> canFrames) {
    final Map<Integer, Long> ids = canFrames.stream().collect(
        Collectors.groupingBy(CANFrame::getId, Collectors.counting()));
    final int uniqueKeys = ids.size();
    log.info("Amount of unique keys: " + uniqueKeys);
    ids.forEach((id, count) ->
        log.info("ID 0x" + Integer.toHexString(id) + ": " + count + " responses"));
  }

  private static void handleResponse(Iterator<CANFrame> responses) {
    while (responses.hasNext()) {
      frames.add(responses.next());
    }
  }

  private static void runFuzzing(final Device device, final API api, final Token token) {
    // open a connection to cloud
    cloud = api.connectToDevice(device, token);
    for (count = 0; count < maxCount; count++) {
      makeRequest(cloud);
    }
    // handle collected frames to handler
    handleData(frames);

    cloud.close();
  }

  private static void start(final String key, final String secret, final String deviceId) {
    // Create API connection class
    final API api = API.get();
    // Get authentication token
    final Token token = api.authenticate(key, secret);
    if (token != null) {
      // Get devices list available for this token
      final Collection<Device> devices = api.getConnectedDevices(token);

      Device device = null;
      for (final Device d : devices) {
        if (deviceId == null ||
            (d.isAvailable() && d.getDeviceId().equalsIgnoreCase(deviceId))) {
          device = d;
          break;
        }
      }

      if (device != null) {
        log.info("Using device: {}", device.getName());
        runFuzzing(device, api, token);
      } else {
        log.error("Couldn't find a device, please use correct device id and/or reserve it");
        die();
      }
    } else {
      log.error("Couldn't authenticate in VT cloud");
    }
  }
}
